File: java/proxy/exercise/VideoServiceProxy.java
package proxy.exercise;

public class VideoServiceProxy implements MediaStreaming {

    private VideoService videoService;
    private User user;

    public VideoServiceProxy(User user) {
        this.videoService = new VideoService();
        this.user = user;
    }

    @Override
    public void playVideo(String videoId) {
        // Authentication and authorization checks
        if (user.hasPermission(videoId)) {
            videoService.playVideo(videoId);
            // Log user interaction
            System.out.println("Logged user interaction: " + user.getUsername() + " watched video " + videoId);
        } else {
            System.out.println("Access denied for video: " + videoId);
        }
    }
}

File: C#/Proxy/Exercise/VideoServiceProxy.cs
namespace Proxy.Exercise;

public sealed class VideoServiceProxy : IMediaStreaming
{
    private readonly VideoService _service;
    private readonly User _user;

    public VideoServiceProxy(User user)
    {
        _service = new VideoService();
        _user = user;
    }

    public void PlayVideo(string videoId)
    {
        if (_user.HasPermission)
        {
            _service.PlayVideo(videoId);
            Console.WriteLine($"Logged user interaction: {_user.Name} watched video {videoId}");
        }
        else
        {
            Console.WriteLine($"Access denied for video: {videoId}");
        }
    }
}
